src/Introspection/IntrospectNullableType.hack

```diff
@@ -0,0 +1,22 @@
+namespace Slack\GraphQL\Introspection;
+
+/**
+* Types are nullable by default in GraphQL, however in this framework, we wrap
+* nullable types with either NullableInputType or NullableOutputType. This trait
+* lets us read introspection fields from the inner type since we don't need to
+* specify anything for nullable fields during introspection.
+*/
+trait IntrospectNullableType implements IHasInnerType {
```

Comment:
TIL traits can implement interfaces.

Reply:
I do wonder why we need both this and `IHasInnerType` though.

Reply:
I needed a way to access `inner_type` on NullableInput and NullableOutput, however both those are different types: `InputType<TInner>` and `OutputType<TInner, TResolved>`.

I tried setting:

```
trait IntrospectNullableType implements IHasInnerType {
    protected __Type $inner_type;
    ...
}
```

But then when we use the trait in the nullable types it complains about a type mismatch. The interface lets both of those methods return inner type as __Type and for us to access that in this trait.

Reply:
Hmm. I think you can make it work with the following:

```hack
trait IntrospectNullableType<+T as __Type> {
    private T $inner_type;

    protected function getInnerType(): __Type {
        return $this->inner_type;
    }

    // etc
}

final class NullableInputType<+TInner as nonnull> extends InputType<?TInner> {
    use GraphQL\Introspection\IntrospectNullableType<InputType<TInner>>;

    // etc
}

final class NullableOutputType<TInner, TResolved> extends OutputType<?TInner, ?TResolved> {
    use GraphQL\Introspection\IntrospectNullableType<OutputType<TInner, TResolved>>;

    // etc
}
```

Reply:
I didn't even think to use generics, great call.

Reply:
You probably don't need generics for introspection types, since we don't do thing like coercion/field resolving that we would want to write in a type-safe way.